index.html

```diff
@@ -0,0 +1,451 @@
+
+```graphql
+#::[Query with resolvers and realized types]
+query {          # resolver                    | realizes
+                 # ----------------------------|--------------
+  rockets {      # rockets::Query.rockets     -> rockets::Rocket
+    captain {    # rockets::Rocket.astronaut  -> rockets::Astronaut
+      name       # astronauts::Astronaut.name -> String
+      rocket {   # rockets::Astronaut.rocket  -> rockets::Rocket
+        id       # rockets::Rocket.id         -> ID
+      }
+    }
+  }
+
+}
+```
+
+These distinctions are immaterial to the client. As far as clients are concerned, all selections on `Rocket` are the same.
+
+This illusion is maintained by the router, which must break incoming queries into subqueries. In the example above, the path `rockets.captain` is resolved by the `rockets::Rocket.astronaut` resolver, so its selection set is realized as a `rockets::Astronaut`. However, there is no `name` field for `rockets::Astronaut`—instead, the only resolver for `Astronaut.name` is `astronauts::Astronaut.name`. To process this query, the router must "convert" a `rockets::Astronaut` to an `astronauts::Astronaut` via an `_entities` query.
+
+### Portability
+A type is *portable to* a subgraph if it has a `@csdl_key` for that subgraph.
+
+### Free / Bound fields
+
+Free fields can be resolved by any subgraph, bound fields can only be resolved by the subgraph they're bound to.
+
+### Mandatory Types
+
+A CSDL document **must** define a `csdl_Graph` enum. Each enum value defines a subgraph. Each value **must** be annotated with a `@csdl_endpoint` directive specifying the endpoint URL for the subgraph.
+
+```graphql
+#::[Using `csdl_Graph` to define endpoints]
+enum csdl_Graph {
+  ROCKETS @csdl_endpoint(url: "https://rockets.example.com"),
+  ASTRONAUTS @csdl_endpoint(url: "https://astronauts.example.com"),
+}
+```
+
+The `csdl_Graph` enum is used as input to csdl directives which link fields and types to subgraphs.
+
+## Provided Types
+
+Valid CSDL documents **must** include the declarations of any CSDL types they use.
+
+### csdl_Never
+```graphql
+#::[Declaration of csdl_Never]
+type csdl_Never {}
+```
+
+The `csdl_Never` type is used to indicate the type of a field which **cannot be resolved**. This is typically due to a composition error. Producers **should** provide more specific error information with a `@csdl_Error` directive on the relevant field.
+
+```graphql
+#::[Using `csdl_Never` and `@csdl_error` to mark a field error]
+type SomeType {
+  someField: csdl_Never
+    @csdl_error(message: "someField defined with conflicting types: Int, String")
+}
+```
+
+## Scalars
+
+### `scalar csdl_SelectionSet`
+
+A GraphQL selection set represented with the same syntax as defined in [the GraphQL SDL](https://spec.graphql.org/draft/#sec-Selection-Sets).
```

Comment:
Are all legal selection set features allowed?

- aliases
- field arguments
- directives on fields
- named fragment spreads (including directives applied to the spread)
- inline fragments (including directives on the inline fragment)

Or do we want to restrict the set of allowed features?

Reply:
All legal selection set features are allowed. Inline fragments are specifically ok and work now (or should, I believe). Named fragment spreads would work, but there's nowhere to declare fragments, so.

Added language to clarify this.

Reply:
Variables in input values wouldn't easily work though, so I think we'd need to explicitly forbid those.

One large pain point we've experienced in our federated backend: arguments that have type `_FieldSet` make it impossible to execute queries at the root level. This is generally fine for `@key` and `@provides`, but has been painful for `@requires`, since often we'd want to require information like auth (with e.g. `query { me { id } }`) but were unable to access that information at the level of the type containing the field with `@requires`. This has led to some horrible hacks in our schema, where we basically take fields that exist at root and then copy-paste their resolver code into a new field at the `@requires` field's type, or just bypass federation entirely and query the subgraph directly. This isn't something I think we have to address right away, but I do think it's something we should eventually address in some version of this spec.

Reply:
yeah, i mean, if i were writing this from whole cloth today, we probably wouldn't have `key` and wouldn't have any reference to a specific `_entities` query but would instead have the ability to specify (bulk and non-bulk) query partials to port from one service to another.

but i think that's right at the point where it'll be a pretty significant change to the query planner, which i don't think we want to block on. how significant a change is this? idk, the `_entities` query is basically hardcoded right now. it might not be too much work to use a particular `ExecutableDefinition` as a template instead, but I could also see that change being bigger than we'd expect

Reply:
Just wanted to flag https://github.com/apollographql/federation/issues/272#issuecomment-742228345 here for visibility. Short version is that I don't really see the benefit of requiring braces and for consistency with user-visible syntax it might be better to stick with `fields`.

Reply:
I definitely agree the kind of use cases @sachindshinde describes need a better solution. I don't think being able to specify arbitrary queries in `@requires` would be the answer though. `@requires` is specifically meant to add additional fields to the representation of an entity that gets passed into a service. Although `_entities` is a weird hack, and we pass those representations in as arbitrary data, conceptually they are typed, and they are meant to correspond to what a parent resolver would return.

Taking that correspondence further, I think what you want here is not a way to pass in additional data as a `parent` value, but as `context`. I suspect that also fits better with common use cases, because `query { me { id } }` is not something you want to pass just to a single resolver, but is context you's want passed into a service and made available to every resolver. Just brainstorming here, but I can see us allowing arbitrary fragments to be specified as context a particular service requires. That would make the gateway responsible for fetching that data first (if it doesn't have it already), and then passing it through on `extensions` in the request.

Reply:
in the current draft, keys are defined as fragments on types, which i think resolves these q's.

agree that `@requires` should probably not just throw arbitrary stuff into the `representation`. but i'd also avoid having service-global "context queries", but keep those granular at the field level (but probably provided through extensions or similar)

Reply:
Keys as fragments on types still leaves these questions for `@provides` and `@requires` I think. We could go with fragments there as well, which seems to make semantic sense. But while`@cs_key` gives us an easy way to bind a fragment to a graph, that won't work here because `@provides` and `@requires` are defined on specific fields. That suggests we'd need to specify fragment names in `requires` and `provides` as part of a `@cs_resolve`. I'm not thrilled by a proliferation of generated fragment names and the extra level of indirection that means you can't look at a `@cs_resolve` and see what it requires and/or provides. But maybe that doesn't really matter because CSDL isn't meant to be consumed by humans?